File: Comprehensive_Data_Operations_Simulation/Q1. Range Module/Q1_RangeModule.py
class RangeModule(object):

    def __init__(self):
        self.intervals = []  # list to store tracked ranges as [start, end)

    def addRange(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: None
        """
        new_intervals = []   # to build the updated interval list
        placed = False       # tells if merged interval is already added

        for l, r in self.intervals:  # check each stored interval
            if r < left:
                # current interval lies completely before new range
                new_intervals.append([l, r])
            elif right < l:
                # new range lies before this interval
                if not placed:
                    new_intervals.append([left, right])  # add merged new range
                    placed = True
                new_intervals.append([l, r])  # add current interval
            else:
                # intervals overlap → merge them
                left = min(left, l)
                right = max(right, r)

        if not placed:
            # if merged interval not added yet, add at the end
            new_intervals.append([left, right])

        self.intervals = new_intervals  # update intervals list

    def queryRange(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: bool
        """
        for l, r in self.intervals:  # check each stored interval
            if l <= left and r >= right:
                # if interval fully covers [left, right)
                return True
        return False  # not fully covered

    def removeRange(self, left, right):
        """
        :type left: int
        :type right: int
        :rtype: None
        """
        new_intervals = []   # new interval list

        for l, r in self.intervals:  # check each stored interval
            if r <= left or l >= right:
                # interval is completely outside the removed range
                new_intervals.append([l, r])
            else:
                # overlapping part to remove
                if l < left:
                    new_intervals.append([l, left])  # keep left part
                if r > right:
                    new_intervals.append([right, r]) # keep right part

        self.intervals = new_intervals  # update intervals

File: Comprehensive_Data_Operations_Simulation/Q1. Range Module/Q1_RangeModule.java
import java.util.*;

class RangeModule {

    TreeMap<Integer, Integer> map;   // stores intervals as: start -> end

    public RangeModule() {
        map = new TreeMap<>();       // create the TreeMap
    }
    
    public void addRange(int left, int right) {

        Integer start = map.floorKey(left);  // find interval that starts before or at 'left'

        if (start != null && map.get(start) >= left) {   // if it overlaps the new range
            left = Math.min(left, start);                // expand left boundary
            right = Math.max(right, map.get(start));     // expand right boundary
            map.remove(start);                           // remove the old interval
        }

        Integer next = map.ceilingKey(left);             // find next interval starting after 'left'
        while (next != null && next <= right) {          // while intervals overlap
            right = Math.max(right, map.get(next));      // merge right boundary
            map.remove(next);                            // remove merged interval
            next = map.ceilingKey(left);                 // go to next
        }

        map.put(left, right);                            // add final merged interval
    }
    
    public boolean queryRange(int left, int right) {

        Integer start = map.floorKey(left);              // find interval starting before 'left'

        if (start == null) return false;                 // if no such interval, range not covered

        return map.get(start) >= right;                  // check if interval fully covers [left, right)
    }
    
    public void removeRange(int left, int right) {

        Integer start = map.floorKey(left);              // find interval < left

        if (start != null && map.get(start) > left) {    // if it overlaps the left boundary
            int end = map.get(start);                    // get its end
            map.remove(start);                           // remove original interval

            if (start < left) map.put(start, left);      // left portion remains
            if (end > right) map.put(right, end);        // right portion remains
        }

        Integer next = map.ceilingKey(left);             // find next interval to check
        while (next != null && next < right) {           // while inside removal area
            int end = map.get(next);                     // get its end
            map.remove(next);                            // remove interval

            if (end > right) {                           // if interval extends beyond 'right'
                map.put(right, end);                     // add back right part
                break;                                   // stop processing
            }

            next = map.ceilingKey(left);                 // move to next
        }
    }
}

/**
 * Your RangeModule object will be instantiated and called as such:
 * RangeModule obj = new RangeModule();
 * obj.addRange(left,right);
 * boolean param_2 = obj.queryRange(left,right);
 * obj.removeRange(left,right);
 */
